posts/existential-haskell.org

```diff
@@ -0,0 +1,319 @@
+Most of the world’s statically-typed programming languages allow their users to write code resembling the following Java:
+
+#+BEGIN_SRC java
+  public static Comparable someFn() {
+      return "a concrete String value";
+  }
+#+END_SRC
+
+Syntactically, this code is uncontroversial: it’s a function that returns a value. Its only interesting aspect lies in the function signature−even though the function body returns a value of type ~String~, its return type is declared to be [[https://docs.oracle.com/javase/8/docs/api/java/lang/Comparable.html][~Comparable~]], which is not a concrete data type, but a Java [[https://en.wikipedia.org/wiki/Interface_(computing)][interface]]. We can try to write the same thing in Haskell:
+
+#+BEGIN_SRC haskell
+  someComparableValue :: Ord a => a
+  someComparableValue _ = "a concrete string value"
+#+END_SRC
+
+Because this is not semantically-valid Haskell, we get the following error:
+
+#+begin_src
+<interactive>:3:27: error:
+    • Couldn't match expected type ‘a’ with actual type ‘[Char]’
+      ‘a’ is a rigid type variable bound by
+        the type signature for:
+          someComparableValue :: forall a. Ord a => Int -> a
+#+end_src
+
+Haskell’s typechecker looks at the body of this function and says “hey, man, you’re returning a concrete string value here, not ’any type that is ~Ord~--erable.’” Though this is a valid notion in Java, it’s not valid in Haskell. Another perspective on this is that Java allows a value to have more than one type: we can treat a Java string literal as a value of type ~java.lang.String~, or of type[fn:2] ~Comparable~, or of its superclass ~java.lang.Object~. This application of the [[https://en.wikipedia.org/wiki/Rule_of_least_power][rule of least power]] is a useful one: sometimes we want to hide the implementation details of a function’s return types. Haskell, being a powerful language, supports this use case with what are called /existential/ types.
+
+What follows is an overview of existential types in practice, and the quirks associated therewith. I write this not because I expect to break any new ground—all the techniques I use here are long-documented in the literature, and Haskell veterans will probably find little new in this post[fn:0]—but because the existing resources are scattered, perhaps oddly so given how central dynamic dispatch is to most programming languages that aren't Haskell, and because exploring the edge cases in the design illustrates the compromises inherent in language and library design.
+
```

Comment:
I love this and would kinda prefer it upfront? I like how you hook readers in with the comparison to Java being loosey goosey with interfaces vs. concrete types, but as a reader I prefer to know upfront what I'm getting into. Another idea is to break up the above section with more sub-headings.

This is also biased by the accessibility required for my brain, where I need to constantly remind myself what I am doing, what I am reading, to make sure I am tracking, so if that suggestion totally kills your vibe and art, I respect that.

Reply:
Hmm, let me think about this. I think I like the hook, but maybe I can split it into two parts, or a summary header.